src/core/inductive.ml

```diff
@@ -0,0 +1,213 @@
+  in
+
+  (* STEP 1: Create the type of the property p *)
+  let prop = _Symb c.symb_Prop in
+  let codom ts = _Impl (fapp ind_sym ts) prop in
+  let p_type = gen_ind_typ_codom pos ind_sym codom in
+
+  (* STEP 2: Create each clause according to a constructor *)
+  (* [case_of cons_sym] creates a clause according to a constructor
+     [cons_sym]. *)
+  let case_of : sym -> tbox = fun cons_sym ->
+    let codom : term list -> tvar list -> tbox -> tbox = fun ts xs _ ->
+      prf_of_p (List.map lift ts)
+        (app (_Symb cons_sym) (List.rev_map _Vari xs))
+    in
+    let get_name : tbinder -> int -> tvar * term = fun b _ ->
+      Basics.unbind_name b
+    in
+    let build_rec_hyp : term list -> tvar -> tbox = fun ts x ->
+      prf_of_p (List.map lift ts) (_Vari x)
+    in
+    let domrec : term -> tvar -> tbox -> tbox -> tbox =
+      fun a x rec_hyp next ->
+      _Prod (lift a) (Bindlib.bind_var x (_Impl rec_hyp next))
+    in
+    let dom : term -> tvar -> tbox -> tbox = fun a x next ->
+      _Prod (lift a) (Bindlib.bind_var x next)
+    in
+    let f_rec : tbox -> tvar -> tbox -> tbox = fun a _ _ -> a in
+    let f_not_rec : tbox -> tvar -> tbox = fun a _ -> a in
+    let acc : tbox = _Type in
+    fold_cons_typ pos codom get_name build_rec_hyp domrec dom ind_sym
+      cons_sym f_rec f_not_rec acc
+  in
+
+  (* STEP 3: Create the conclusion of the inductive principle *)
+  let codom ts =
+    let x = Bindlib.new_var mkfree "x" in
+    let t = Bindlib.bind_var x (prf_of_p ts (_Vari x)) in
+    _Prod (fapp ind_sym ts) t
+  in
+  let conclusion = gen_ind_typ_codom pos ind_sym codom in
+
+  (* STEP 4: Create the induction principle *)
+  let t =
+    List.fold_right (fun a b -> _Impl (case_of a) b) cons_list conclusion
+  in
+  let t = _Prod p_type (Bindlib.bind_var p t) in
+  if !log_enabled then
+    log_ind "The induction principle of the inductive type [%a] is %a"
+      Pretty.pp_ident (Pos.none ("ind_"^ind_sym.sym_name))
+      Print.pp_term (Bindlib.unbox t);
+  Bindlib.unbox t
+
+(** [gen_rec_rules ind_sym rec_sym cons_list] returns the p_rules associated
+    with an induction principle [rec_sym] of the inductive type [ind_sym]
+    (with its constructors [cons_list]). That means the command
+    inductive T : Π(x1:A1),...,Π(xm:Am), TYPE := c1:T1 | ... | cn:Tn generates
+    a rule for each constructor. If Ti = Π(b1:B1),...,Π(bk:Bk), T then the
+    rule for ci is:
+    rule ind_T p pi1 ... pin _ ... _ (ci x1 ... xk) --> pii x1 t1? ... xk tk?
+    with m underscores and tj? = nothing if xj isn't a value of the inductive
+    type and tj? = (ind_T p pi1 ... pin _ ... _ xj) if xj is a value of the
+    inductive type T (i.e. xj = T v1 ... vm) *)
+let gen_rec_rules : sym -> sym -> sym list -> p_rule list =
+  fun ind_sym rec_sym cons_list ->
+
+  (* STEP 1: Create the common head of the rules *)
+  let rec_iden = P.iden rec_sym.sym_name in
+  let p_patt = P.patt0 "p" in
+  let add_arg t s = P.appl t (P.patt0 ("pi" ^ s.sym_name)) in
```

Comment:
To be fully correct, we should make sure that no other variable in the rule can have this name. Is it the case?

Reply:
I think no (an argument of constructor "c" could be name "pic"!). That's why it is better to use names like p1, p2, x1, x2, etc. to make sure that there is no name clashes.

Reply:
There is no check, you right. But there is no clash too because each constructor argument has the name xi and each constructor has a different name.

Reply:
I can change this but there is a clash the GIP however, I think.

Reply:
How is it done that constructor arguments are named xi?

Reply:
```
let get_name : tbinder -> int -> p_term * term = fun b i ->
      let (_,b) = Basics.unbind_name b in
      (P.patt0 ("x" ^ (string_of_int i)), b)
    in
```
here :)

Reply:
Il faut documenter cela dans le code car c'est un point important.

Reply:
Ah yes it's not clear because it's the same name in the quantification you right!

Reply:
This problem is still not fixed in the current code. The use of explicit names like "x", "p", "pi" should be avoided. We should use get_safe_prefix instead like in gen_rec_type.

Remark: in fact, there was no possible name clashes in gen_rec_type as gen_rec_type works on tbox. On the other hand, there are possible name clashes in gen_rec_rules and gen_rule_cons as these functions work on p_term.

Reply:
Ok, I was wrong. The point is that there cannot be name clashes between pattern variables and function symbols since pattern variables are prefixed with `$`.